lib/plugins/aws/deploy/lib/checkForChanges.test.js

```diff
@@ -655,6 +655,53 @@ describe('checkForChanges', () => {
         expect(awsDeploy.serverless.service.provider.shouldNotDeploy).to.equal(undefined);
       });
     });
+
+    it('should set a flag if the remote and local hashes are same with package.artifact in .serverless directory', () => {
+      awsDeploy.serverless.service.package = {
```

Comment:
We stay away from introducing tests like that, as they too much attached to internals and due that they're painful for any refactors.

Ideally all new tests should be based on `runServerless` util. See https://github.com/serverless/serverless/tree/master/test#unit-tests

In this case I think we need to prepare a dedicated fixture, issue `runServerless` against it with `['package']` args, and confirm on the outcome that given issue is not observed

Reply:
I've been working at this for a while -- what I'm a bit confused on is how to have a mocked FS with `.serverless/my-own.zip` that returns a hash of my choosing. Any tips?

Reply:
One other thought I had in introducing this new test is that the unit tests for `checkForChanges` aside from the single `runServerless`-based tests run in virtually no time at all. By contrast, the `runServerless`-based `Should recognize package.artifact` test takes 440ms on my machine.

At a meta level, I'm definitely a huge proponent of integration / semi-integration tests like this, but judiciously pared so as to keep the entire test suite blazingly fast (I even gave a talk on the specific subject years back 😄 ). My attempt so far with a `runServerless`-based test that performs real file I/O is on par with the other one at around 400-500ms. For a single line code change, this seems like it might not a sustainable path to have perhaps _all changes_ go through `runServerless` as I can see the overall test suite time taking forever quickly if a change this size alone adds a half second to the suite...

I'm fine with whatever to ship the change, and am just trying to be mindful of the overall DX for the core team as well as outside contributors.

Reply:
> Do you have any good examples of creating a fixture wherein .serverless can exist and be pre-populated?

I think we should simply add (commit in) `.serverless` folder with some artifact zip, and `serverless.yml` should reference this file

We already have other fixtures with artifacts committed in, e.g.: https://github.com/serverless/serverless/tree/master/test/fixtures/packageArtifact

> By contrast, the runServerless-based Should recognize package.artifact test takes 440ms on my machine.

Indeed `runServerless` tests will definitely be more time taking. Still we already have a lot of them, and total test time seems acceptable. We also optimize things by using as least`runServerless` runs as possible (various cases in one file are often tested via one service run).

It might also be the case, that for some reason your `runServerless` test case runs significantly slower than other `runServelress` tests as found in so far tests. If you observe that, I'll further investigate why it's the case

Reply:
On the side note, many contributors already contritbuted `runServerless` tests, and so far feedback was just positive (in many cases it helps to not write painful stubs). It's first time, where performance issue was reported, so I would double check if indeed all is good with setup test.

Reply:
Thanks for the responses! I'm up late so I might be able to catch you for some more input your time...

**Slow Tests**

I'm on an old MacBook 12". Definitely not a fast machine. The slowness I'm seeing happens on the existing test I'm modeling:

```
  checkForChanges #2
    ✓ Should recognize package.artifact (640ms)
```

Thinking about this more, this particular test has to (1) file I/O to read the zips and (2) do crypto hashing. This might not be the case for many/any other of the `runServerless` tests, but my guess is that you just can't get file I/O + crypto work done anywhere in the ballpark as fast as stubbed out stuff.

**My new test**

So, I already was modeling on `packageArtifact`. The problem is that if I have `test/fixtures/packageArtifactInServerlessDir/my-own.zip` and reference `package.artifact: my-own.zip` everything works. But for my case I need to have `test/fixtures/packageArtifactInServerlessDir/.serverless/my-own.zip` `package.artifact: .serverless/my-own.zip` and I instead get:

```
  1) checkForChanges #2
       should skip a deployment with identical hashes and package.artifact targeting .serverless directory:
     Error: Could not add file content to hash: Error: ENOENT: no such file or directory, open '.serverless/my-own.zip'
      at ReadStream.<anonymous> (lib/plugins/aws/package/compile/functions/index.js:681:16)
      at ReadStream.EventEmitter.emit (domain.js:483:12)
      at internal/fs/streams.js:136:12
      at FSReqCallback.oncomplete (fs.js:156:23)
  From previous event:
      at AwsCompileFunctions.compileFunction (lib/plugins/aws/package/compile/functions/index.js:103:25)
      at /Users/rye/scm/fmd/serverless/lib/plugins/aws/package/compile/functions/index.js:623:62
  From previous event:
      at AwsCompileFunctions.compileFunctions (lib/plugins/aws/package/compile/functions/index.js:623:22)
  From previous event:
      at Object.package:compileFunctions [as hook] (lib/plugins/aws/package/compile/functions/index.js:32:12)
  From previous event:
      at PluginManager.invoke (lib/classes/PluginManager.js:510:22)
      at PluginManager.spawn (lib/classes/PluginManager.js:530:17)
      at /Users/rye/scm/fmd/serverless/lib/plugins/deploy/deploy.js:122:50
  From previous event:
      at Object.before:deploy:deploy [as hook] (lib/plugins/deploy/deploy.js:102:22)
      at /Users/rye/scm/fmd/serverless/lib/classes/PluginManager.js:510:55
  From previous event:
      at PluginManager.invoke (lib/classes/PluginManager.js:510:22)
      at /Users/rye/scm/fmd/serverless/lib/classes/PluginManager.js:545:24
  From previous event:
      at PluginManager.run (lib/classes/PluginManager.js:545:8)
      at /Users/rye/scm/fmd/serverless/lib/Serverless.js:168:33
  From previous event:
      at Serverless.run (lib/Serverless.js:155:74)
      at /Users/rye/scm/fmd/serverless/node_modules/@serverless/test/run-serverless.js:268:30
      at processImmediate (internal/timers.js:456:21)
  From previous event:
      at /Users/rye/scm/fmd/serverless/node_modules/@serverless/test/run-serverless.js:179:50
```

You can see the current WIP code and test in a temp branch at: https://github.com/serverless/serverless/compare/master...FormidableLabs:bug-zip-hash-check-new-tests and see the test failure as I have above with:

```sh
$ npm test
```

(I've `it.only()`-ed my WIP test).

Any guidance for getting the above to work (or alternate approaches) would be most appreciated!